Add support for Lambda@Edge

pmuens opened the pull request:
This is a rebased version of https://github.com/serverless/serverless/pull/4796 from @laardee 👍 🥇 🙌

## What did you implement:

Closes #3944 

Adds support for Lambda@Edge via pure CloudFormation. Thanks again to @laardee who jumped in to work on an initial implementation for this!

## How did you implement it:

Compile all the necessary CloudFormation resources (CloudFront / Versioned Lambda function, etc.) to add support for Lambda@Edge.

## Todos:

- [x] Retain Lambda Functions to avoid errors on Stack removals
- [x] Log info that functions have to be removed by hand when removing stack

## How can we verify it:

Create a new serverless service and replace the `handler.js` and `serverless.yml` code with the following snippets:

```javascript
// handler.js
module.exports.cloudfront = (event, context, callback) => {
  const response = event.Records[0].cf.response;
  const headers = response.headers;

  headers['x-serverless-time'] = [{ key: 'x-serverless-time', value: Date.now().toString() }];

  return callback(null, response);
};
```

```yaml
service: test-${self:custom.idx}

provider:
  name: aws
  runtime: nodejs10.x
  versionFunctions: false

custom:
  idx: 0

functions:
  hello:
    handler: handler.hello
    events:
      - cloudFront:
          eventType: viewer-response
          origin: https://serverless.com/framework/docs
```

Deploy the service (this could take a couple of minutes (~10 - 20mins)). After that you can access the CloudFront distribution URL which is printed in your console (you can also use `serverless info`). This should open up the Serverless Framework docs page.

Inspect the headers (e.g. using [this guide](https://www.mkyong.com/computer-tips/how-to-view-http-headers-in-google-chrome/) if you're running Chrome). You should see the `x-serverless-time` header which was added by the Lambda@Edge function.

Once validated you might want to remove the stack. Note the info message about the necessary manual Lambda@Edge function removals which is logged on the console.

## Todos:

_**Note: Run `npm run test-ci` to run all validation checks on proposed changes**_

- [x] Write tests and confirm existing functionality is not broken.  
       **Validate via `npm test`**
- [x] Write documentation
- [x] Ensure there are no lint errors.  
       **Validate via `npm run lint-updated`**  
       _Note: Some reported issues can be automatically fixed by running `npm run lint:fix`_
- [x] Ensure introduced changes match Prettier formatting.  
       **Validate via `npm run prettier-check-updated`**  
       _Note: All reported issues can be automatically fixed by running `npm run prettify-updated`_
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

pmuens commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -0,0 +1,280 @@
+                  Ref: this.provider.naming.getLambdaLogicalId(functionName),
+                },
+              },
+            });
```
Good idea. I'll keep it as is since the current implementation is more readable.

pmuens commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -0,0 +1,280 @@
+
+      _.merge(Resources, { [this.cloudFrontDistributionLogicalId]: CloudFrontDistribution });
+
+      _.merge(Outputs, {
```
Had to keep `_.merge` here. Using `Object.assign` caused the tests to fail.

pmuens commented:
> Some of tests crash when run individually (exposed by isolated test run)

Good catch. Just pushed a fix...

medikoo commented on lib/plugins/aws/lib/monitorStack.js:
```diff
@@ -5,19 +5,27 @@ const async = require('async');
     const cfQueryString = `region=${region}#/stack/detail?stackId=${encodedStackId}`;
     const stackUrl = `${baseCfUrl}?${cfQueryString}`;
 
+    if (options.async) {
```
So we still need have that issue with`Retain` in?

Is the deployments of those lambdas that slow? If we would do it via SDK, will we face the same problem?

Still, even it appears that we can't escape long deployments. I think not waiting for result is pretty controversial, as stack fails are not reported to the user.

Introducing that, we force the user to manually confirm with AWS console that stack deployment was successful (which is not very DX friendly). I think it'll be better to leave this long going process running, then user easily can confirm on a progress in a terminal. I would just output meaningful log as:

```
This deployment may take significant amount of time due to {reasons}.
If you want, you may stop monitoring of it with Ctrl+C (it won't stop the deployment on AWS side), and confirm later at {link} whether deployment finished successfully.
Note that no other deploy of this stack can be started until ongoing one finalizes.
```

Additional drawback of going _async_ is that going _async_ way, may leave impression that user can go on with updates and re-deploy shortly, which is not true. So I think long-going process makes a valuable message.

As a side note, an _async_ deployments make a nice use case for eventual native menu bar app, as it's the one that can monitor in background CF deployments and issue notifications when one finalizes (having that, `async` option will become more valid I think)

pmuens replied:
Good points and I agree with most of them. I talked with @ganeshvlrk about this and we thought that it might be a good idea to do deployments and removals async since it takes on avg. ~20 minutes for a deployment to finish.

Ganesh had the idea that we can add a flag to await the deployment. Maybe we can invert this and provide a flag to do deployments async. This can be combined with your log message output proposal above.

> If we would do it via SDK, will we face the same problem?

I looked into support for Lambda@Edge via custom resources the other day and my conclusion was that it's not worth it since we would be dealing with the same drawbacks we're facing with CloudFormation deployments / removal.

> As a side note, an async deployments make a nice use case for eventual native menu bar app, as it's the one that can monitor in background CF deployments and issue notifications when one finalizes (having that, async option will become more valid I think)

I like that idea and i had a similar concept in my head when we worked on the EventGateway back then, but that's for another day... Although throwing an Electron app together which targets the use case you mentioned might be pretty simple to do.

medikoo replied:
> Maybe we can invert this and provide a flag to do deployments async. This can be combined with your log message output proposal above.

I wonder whether in first place it'll be better to add support for handling Ctrl+C gently (and in certain (or all) cases suggesting user can stop monitoring that way)

As that way user will be definitely more aware of nature of the action.

In Node.js we can attach listener for exists done via Ctrl+C way.. so we probably can handle that nicely (like: abort naturally and display nice message with link where stack can be monitored). Not sure though if same can be done on Windows (that probably should be investigated before we go that route)

Then, still we may have use case for forcing that via config (e.g. user in all cases wants to abort early, so prefers it's just done for him).

I propose that we follow that route:

- Make this PR strictly about Lambda@Edge support (propagate eventual async support to another PR)
- (to be decided) In other PR if that's proven possible and reliable, introduce a way of gentle Ctrl+C handling
- (to be decided) In other PR, if we agree it's worthwhile, introduce an option that will generally allow _async_ deployments.

pmuens replied:
Sounds like a good plan. I'll update this PR 👍

medikoo commented on lib/plugins/aws/lib/naming.test.js:
```diff
@@ -811,4 +811,34 @@ describe('#naming()', () => {
   });
+
+  describe('#getCloudFrontDistributionLogicalId()', () => {
+    it('should return CloudFront distribution logical id', () => {
```
I wonder whether there's a real value in such unit tests (?)

Technically we write same code twice (it's a bit as writing a function in one file and in other confirming that code of the function matches given string - and that feels counterproductive and not really valuable).

In case of naming I would rather rely on tests of given functionality (which relies on `naming`). Over there via spies we can confirm that given functionality relies on `naming` (but without inspecting what naming function returns)

pmuens replied:
Good point. I agree with the sentiment :+1:

Not sure if we always test for the correct naming generation in the other tests though (what the coverage of the `naming.js` code is if we remove the tests / test file).

medikoo replied:
> Not sure if we always test for the correct naming generation in the other tests though (what the coverage of the naming.js code is if we remove the tests / test file).

Yes, I don't think we should call for any refactor now. But we may just add here special note (for future PR's), and stop writing more tests here, as in most cases they miss the point.

pmuens commented on docs/providers/aws/events/cloudfront.md:
```diff
@@ -0,0 +1,161 @@
+- `origin-response`, when CloudFront receives the response from the origin service
+- `viewer-response`, before the response returned to the client
+
+**IMPORTANT** Due to current [Lambda@Edge limitations](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/lambda-edge-delete-replicas.html) it's necessary to set the `DeletionPolicy` to `Retain` for AWS Lambda functions which use the `cloudFront` event. The Serverless Framework will do this automatically for you. However bear in mind that **you have to delete those AWS Lambda func ...
```
Yes, good question :+1:

I tested it and AWS won't remove the Lambda once the CloudFront distribution is removed. I guess that's because the AWS Lambda function is still its own entity compared to the replicas which are bound to the Lambda in a 1:n relationship.

Replicas will be automatically removed and once that happened, you can remove the Lambda (you can't immediately remove the Lambda once the stack removal was successful).

medikoo replied:
Ok, thanks for clarification.

Indeed currently we don't have means to address  that cleanly.

Still if at some point we will have native menu bar app, then deletion of that lambda might be delegated to it (another nice use case for an app)

pmuens replied:
Sounds good! :+1:

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -0,0 +1,305 @@
+        "Don't forget to manually remove your Lambda@Edge functions ",
+        'once the CloudFront distribution removal is successfully propagated!',
+      ].join('');
+      this.serverless.cli.log(message, 'Serverless', { color: 'orange' });
```
If I see correctly, it's output also if there are no ClouFront endpoints involved (?)

To avoid unnecessary noise, it'll be cleaner to output that only if there were any CF endpoints in a stack.

Really solid would be if we also warn if single CF  endoint is removed (so also in special update cases). Still I imagine it'll be more time-taking, and I guess it'll be worth to come up with some generic handler (so other plugins can benefit from such inspection as well). So if at all, I would leave it for other PR

pmuens replied:
Yes, that was the initial intention. Good catch. Will update in a minute...

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -0,0 +1,305 @@
+        // ensure that functions are versioned
+        Object.assign(functionObj, { versionFunction: true });
+        // set the maximum memory size if not explicitly configured
+        if (!functionObj.memorySize) {
```
I guess it'll be nice to introduce a user friendly validation here. So we crash with meaningful message (before deployment) stating that configured `memorySize` or `timeout` is not valid for CF endpoint case

pmuens replied:
Sounds like a good idea... Just pushed an updated version.

medikoo commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -0,0 +1,280 @@
+
+            let origin = createOrigin(functionName, event.cloudFront.origin, this.provider.naming);
+
+            const existingOrigin = _.find(origins, o => o.OriginPath === origin.OriginPath);
```
Seems missed (?)

pmuens commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -0,0 +1,336 @@
+          .events.find(e => Object.keys(e)[0] === 'cloudFront')
+      );
+    }
+    if (this.serverless.processedInput.commands[0] === 'remove' && isEventUsed) {
```
Hmm. Doesn't JavaScript support lazy evaluation? I wrote it in that way so that the find is only fired when the command is `remove`.

pmuens replied:
I see what you're saying... This is also the reason why our tests are currently failing.

Working on a fix right now...

pmuens commented:
Just did another rebase and fixed the issue where the removal log info is always executed. It's now only shown when the command is `remove`.

Furthermore I did a deep dive into the testing today and could validate that it's working as expected. Shoutouts to @laardee who helped me here and provided https://github.com/laardee/serverless-lambda-at-edge as an example service 🙇 👍

I also updated the PR description to provide a step-by-step walkthrough to validate that this integration works.

pmuens commented on lib/plugins/aws/package/compile/events/cloudFront/index.js:
```diff
@@ -0,0 +1,336 @@
+    const funcKeys = this.serverless.service.getAllFunctions();
+    let isEventUsed = false;
+    if (funcKeys.length) {
+      isEventUsed = !!funcKeys.find(funcKey =>
```
Good tip, forgot about that. Thanks :+1:

Will merge this now to avoid yet another round of reviews. Maybe we can refactor such usages in the future...

laardee commented:
FINALLY! 🎉:muscle:
